pkg/controllers/tlspolicy/tlspolicy_controller.go

```diff
@@ -0,0 +1,269 @@
+	log := r.Logger().WithValues("TLSPolicy", req.NamespacedName)
+	log.Info("Reconciling TLSPolicy")
+	ctx = crlog.IntoContext(ctx, log)
+
+	previous := &v1alpha1.TLSPolicy{}
+	if err := r.Client().Get(ctx, req.NamespacedName, previous); err != nil {
+		if err := client.IgnoreNotFound(err); err == nil {
+			return ctrl.Result{}, nil
+		} else {
+			return ctrl.Result{}, err
+		}
+	}
+
+	tlsPolicy := previous.DeepCopy()
+	log.V(3).Info("TLSPolicyReconciler Reconcile", "tlsPolicy", tlsPolicy, "tlsPolicy.Spec", tlsPolicy.Spec)
+
+	markedForDeletion := tlsPolicy.GetDeletionTimestamp() != nil
+
+	targetNetworkObject, err := r.FetchValidTargetRef(ctx, tlsPolicy.GetTargetRef(), tlsPolicy.Namespace)
+	log.V(3).Info("TLSPolicyReconciler targetNetworkObject", "targetNetworkObject", targetNetworkObject)
+	if err != nil {
+		if !markedForDeletion {
+			if apierrors.IsNotFound(err) {
+				log.V(3).Info("Network object not found. Cleaning up")
+				err := r.deleteResources(ctx, tlsPolicy, nil)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+			return ctrl.Result{}, err
+		}
+		targetNetworkObject = nil // we need the object set to nil when there's an error, otherwise deleting the resources (when marked for deletion) will panic
+	}
+
+	if markedForDeletion {
+		log.V(3).Info("cleaning up tls policy")
+		if controllerutil.ContainsFinalizer(tlsPolicy, TLSPolicyFinalizer) {
+			if err := r.deleteResources(ctx, tlsPolicy, targetNetworkObject); err != nil {
+				return ctrl.Result{}, err
+			}
+			if err := r.RemoveFinalizer(ctx, tlsPolicy, TLSPolicyFinalizer); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// add finalizer to the tlsPolicy
+	if !controllerutil.ContainsFinalizer(tlsPolicy, TLSPolicyFinalizer) {
+		if err := r.AddFinalizer(ctx, tlsPolicy, TLSPolicyFinalizer); client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	specErr := r.reconcileResources(ctx, tlsPolicy, targetNetworkObject)
+
+	newStatus := r.calculateStatus(tlsPolicy, specErr)
+	tlsPolicy.Status = *newStatus
+
+	if !equality.Semantic.DeepEqual(previous.Status, tlsPolicy.Status) {
+		updateErr := r.Client().Status().Update(ctx, tlsPolicy)
+		if updateErr != nil {
+			// Ignore conflicts, resource might just be outdated.
+			if apierrors.IsConflict(updateErr) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, updateErr
+		}
+	}
+
+	if specErr != nil {
+		return ctrl.Result{}, specErr
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *TLSPolicyReconciler) reconcileResources(ctx context.Context, tlsPolicy *v1alpha1.TLSPolicy, targetNetworkObject client.Object) error {
+	// validate
+	err := tlsPolicy.Validate()
+	if err != nil {
+		return err
+	}
+
+	// reconcile based on gateway diffs
+	gatewayDiffObj, err := r.ComputeGatewayDiffs(ctx, tlsPolicy, targetNetworkObject, &TLSPolicyRefsConfig{})
```

Comment:
Does this work for targeted HTTPRoutes as well?

Reply:
Followed the approach we took with DNSPolicy and it only supports targeting a Gateway currently.

The controller itself follows the same pattern that all kuadrant policies use, i believe these can handle HTTPRoutes as targets, but afaik we don't currently.